### PR TITLE
Add Check for Provider Registration and Refactor

### PR DESCRIPTION
**Pull request by jonathan-innis**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/k8s-configuration/azext_k8s_configuration/_utils.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    # Add gitops private key data to protected settings if exists
+    # Dry-run all key types to determine if the private key is in a valid format
+    invalid_rsa_key, invalid_ecc_key, invalid_dsa_key, invalid_ed25519_key = (False, False, False, False)
+    if ssh_private_key_data != '':
+        try:
+            RSA.import_key(_from_base64(ssh_private_key_data))
+        except ValueError:
+            invalid_rsa_key = True
+        try:
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/_utils.py:28]: May be do this only if it was not a valid rsa key, and similarly check for invalid rsa key and invalid ecc key before importing of dsa key, etc.?

File: src/k8s-configuration/azext_k8s_configuration/_utils.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            invalid_dsa_key = True
+        try:
+            key_obj = io.StringIO(_from_base64(ssh_private_key_data).decode('utf-8'))
+            Ed25519Key(file_obj=key_obj)
+        except SSHException:
+            invalid_ed25519_key = True
+
+        if invalid_rsa_key and invalid_ecc_key and invalid_dsa_key and invalid_ed25519_key:
+            raise InvalidArgumentValueError(
+                'Error! ssh private key provided in invalid format',
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/_utils.py:44]: Give the parameter name as is in the error message

File: src/k8s-configuration/azext_k8s_configuration/_validators.py [hunk 3]
```diff
@@ -44,3 +54,63 @@ def __validate_k8s_name(param_value, param_name, max_len):
...
             'Parameter {0} can only contain lowercase alphanumeric characters and hyphens'.format(param_name))
+
+
+def _validate_url_with_params(repository_url, ssh_private_key_set, known_hosts_contents_set, https_auth_set):
+    scheme = urlparse(repository_url).scheme
+
+    if scheme in ('http', 'https'):
+        if ssh_private_key_set:
+            raise MutuallyExclusiveArgumentError(
+                'Error! An ssh private key cannot be used with an http(s) url',
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/_validators.py:65]: Return the parameter name as is in the error message.  Make this change in all error messages

**Review by NarayanThiru — CHANGES_REQUESTED** [review]

File: src/k8s-configuration/azext_k8s_configuration/_utils.py (hunk 1 above)
Comment (jonathan-innis) [reply src/k8s-configuration/azext_k8s_configuration/_utils.py]: Updated with fix

File: src/k8s-configuration/azext_k8s_configuration/_validators.py [hunk 4]
```diff
@@ -44,3 +58,89 @@ def __validate_k8s_name(param_value, param_name, max_len):
         raise InvalidArgumentValueError(
             'Error! Invalid {0}'.format(param_name),
             'Parameter {0} can only contain lowercase alphanumeric characters and hyphens'.format(param_name))
+
+
+def _validate_url_with_params(repository_url, ssh_private_key_set, known_hosts_contents_set, https_auth_set):
+    scheme = urlparse(repository_url).scheme
+
+    if scheme in ('http', 'https'):
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/_validators.py:66]: Should we accept upper case for HTTP/HTTPS?  If not, add validation for that?

File: src/k8s-configuration/azext_k8s_configuration/_validators.py [hunk 5]
```diff
@@ -44,3 +58,89 @@ def __validate_k8s_name(param_value, param_name, max_len):
...
+        except Exception as ex:
+            raise InvalidArgumentValueError(
+                'Error! ssh known_hosts provided in wrong format',
+                'Verify that all lines in the known_hosts contents are provided in a valid sshd(8) format') from ex
+
+
+def _validate_private_key(ssh_private_key_data):
+    invalid_rsa_key, invalid_ecc_key, invalid_dsa_key, invalid_ed25519_key = (False, False, False, False)
+    try:
+        RSA.import_key(_from_base64(ssh_private_key_data))
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/_validators.py:111]: If this is a valid RSA key, can't you just return the control right here?  Why try to import to other keys?

File: src/k8s-configuration/azext_k8s_configuration/_validators.py [hunk 6]
```diff
@@ -44,3 +58,89 @@ def __validate_k8s_name(param_value, param_name, max_len):
...
+            'Verify the key provided is a valid PEM-formatted key of type RSA, ECC, DSA, or Ed25519')
+
+
+# pylint: disable=broad-except
+def _validate_cc_registration(cmd):
+    try:
+        rp_client = _resource_providers_client(cmd.cli_ctx)
+        registration_state = rp_client.get(consts.PROVIDER_NAMESPACE).registration_state
+
+        if registration_state != "Registered":
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/_validators.py:140]: It probably is the same mixed case always; but wondering if it needs to be a case insensitive comparison.  Also, make this a const?

File: src/k8s-configuration/azext_k8s_configuration/custom.py [hunk 7]
```diff
@@ -81,22 +77,25 @@ def create_k8s_configuration(client, resource_group_name, cluster_name, name, re
...
-                             ssh_private_key_set=ssh_private_key_set,
-                             known_hosts_contents_set=known_hosts_contents_set,
-                             https_auth_set=https_auth_set)
+    _validate_url_with_params(repository_url,
+                              ssh_private_key_set=ssh_private_key_set,
+                              known_hosts_contents_set=known_hosts_contents_set,
+                              https_auth_set=https_auth_set)
+
+    # Validate that the subscription is registered to Microsoft.KubernetesConfiguration
+    _validate_cc_registration(cmd)
```
Comment (NarayanThiru) [inline src/k8s-configuration/azext_k8s_configuration/custom.py:98]: I'm curious why we need to do this validation?  Is the default error from ARM not sufficient to explain to the customer?

**Review by NarayanThiru — CHANGES_REQUESTED** [review]

File: src/k8s-configuration/azext_k8s_configuration/custom.py (hunk 7 above)
Comment (jonathan-innis) [reply src/k8s-configuration/azext_k8s_configuration/custom.py:98]: There was discussion among Safeer and the IDC about adding this to give an explicit error back to the user and clear instructions about where to go if this validation fails

**Pull request merged**